### ISSUE TITLE
Make MetricsRegistryImpl more robust

### ISSUE THREAD

**Issue by jmesnil** [smallrye/smallrye-metrics#36]
during code review, we identified several improvements in the MetricsRegistryImpl class:

Method removeMatching is iterating over metricMap. If there is a match remove method is called. This method removes entry from metricMap and from metadataMap.
It seems to be a bat pattern that the entry is removed from the map by external method while the map is currently iterated by the removeMatching method.
I think https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html#remove-- should be called in removeMatching for metricMap entry + invocation of metadataMap.remove(entry.getKey()) + log mesasge;

**Comment by rsvoboda** [comment]
@jmartisk any plans to look into this ?

**Comment by jmartisk** [comment]
If I remember correctly, the point of calling `remove` method for this was that the removal is a relatively complex operation that potentially operates on two maps at once, and most importantly it includes logic to determine whether the `metadataMap` needs to be updated too, so we used it to avoid having to duplicate this logic in multiple methods. If we switched to `iterator.remove()`, we'd have to duplicate a part of the `MetricsRegistryImpl.remove` method and add synchronization.
Certainly it would be possible, but I don't think the code would be nicer, or faster.
But if you guys think it's better and someone sends a PR, I will probably not object, it doesn't really matter that much :)
